From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Sevenrock <sevenrock@hotmail.de>
Date: Mon, 8 Apr 2024 21:05:33 +0200
Subject: [PATCH] drivers: support building moto modules, part II

This further integrates Motorola drivers into the kernel build system.

Mostly copied from
https://github.com/moto-common/android_kernel_motorola_msm-5.4/commits/14/
https://github.com/LineageOS/android_kernel_motorola_sm7325

Signed-off-by: Sevenrock <sevenrock@hotmail.de>
---
 arch/arm64/Kconfig                            |  6 ++
 drivers/Kconfig                               |  4 +
 drivers/Makefile                              |  2 +
 drivers/backlight/Kconfig                     |  3 +
 drivers/backlight/Makefile                    |  3 +
 drivers/backlight/aw99703/Kbuild              |  2 +-
 drivers/backlight/aw99703/Kconfig             |  9 +++
 drivers/backlight/lm3697/Kbuild               |  2 +-
 drivers/backlight/lm3697/Kconfig              |  9 +++
 drivers/backlight/sm5350/Kbuild               |  2 +-
 drivers/backlight/sm5350/Kconfig              |  6 ++
 drivers/fm/Kbuild                             |  5 +-
 drivers/fm/Kconfig                            |  4 +
 drivers/input/misc/Kconfig                    |  4 +
 drivers/input/misc/Makefile                   | 11 ++-
 drivers/input/misc/focal_fps_mmi/Kbuild       | 50 +-----------
 drivers/input/misc/focal_fps_mmi/Kconfig      |  6 ++
 drivers/input/misc/qpnp_power_on_mmi/Kbuild   |  2 +-
 drivers/input/misc/qpnp_power_on_mmi/Kconfig  |  8 ++
 drivers/input/misc/rbs_fps_mmi/Kbuild         | 14 +---
 drivers/input/misc/rbs_fps_mmi/Kconfig        | 25 ++++++
 drivers/input/touchscreen/Kconfig             |  2 +
 drivers/input/touchscreen/Makefile            |  1 +
 .../touchscreen/goodix_berlin_mmi/Kbuild      | 76 +++++-------------
 .../touchscreen/goodix_berlin_mmi/Kconfig     | 19 +++++
 drivers/leds/Kbuild                           |  6 --
 drivers/leds/trigger/Kbuild                   |  5 --
 drivers/leds/trigger/Kconfig                  |  7 ++
 drivers/leds/trigger/Makefile                 | 28 ++++---
 drivers/misc/Kconfig                          |  1 +
 drivers/misc/Makefile                         | 80 ++++++++++++++++---
 drivers/misc/awinic/Kconfig                   |  2 +-
 drivers/misc/ldo_vibrator_mmi/Kconfig         |  3 +
 drivers/mmi_annotate/Kbuild                   |  6 +-
 drivers/mmi_info/Kbuild                       |  8 +-
 drivers/mmi_relay/Kbuild                      |  6 +-
 drivers/moto_f_usbnet/Kbuild                  |  2 +-
 drivers/nfc/samsung/Kbuild                    |  8 ++
 drivers/power/Kconfig                         | 10 +++
 drivers/power/Makefile                        | 17 +++-
 drivers/power/bq2589x_chg_mmi/Kbuild          | 21 +----
 drivers/power/bq2589x_chg_mmi/Kconfig         | 31 +++++++
 drivers/power/bq2597x_mmi/Kbuild              |  7 +-
 drivers/power/bq2597x_mmi/Kconfig             | 16 ++++
 drivers/power/bq2597x_mmi_iio/Kbuild          |  8 +-
 drivers/power/bq2597x_mmi_iio/Kconfig         |  9 +++
 drivers/power/cw2217b_fg_mmi/Kbuild           |  9 +--
 drivers/power/cw2217b_fg_mmi/Kconfig          | 20 +++++
 drivers/power/mmi-smbcharger-iio/Kconfig      |  9 +++
 drivers/power/mmi_charger/Kbuild              |  2 +-
 drivers/power/mmi_charger/Kconfig             | 30 ++++---
 drivers/power/mmi_discrete_charger/Kbuild     | 14 +---
 drivers/power/mmi_discrete_charger/Kconfig    | 24 ++++++
 .../power/mmi_discrete_turbo_charger/Kbuild   | 16 +---
 .../power/mmi_discrete_turbo_charger/Kconfig  | 12 +++
 .../power/mmi_parallel_charger_iio/Kconfig    |  9 +++
 drivers/power/mmi_sc8549/Kconfig              |  9 +++
 drivers/power/qpnp_adaptive_charge/Kbuild     |  6 +-
 drivers/power/rt9426a_fg_mmi/Kconfig          |  9 +++
 drivers/power/sd77426_fg_mmi/Kconfig          |  9 +++
 drivers/power/sgm4154x_chg_mmi/Kbuild         | 26 +-----
 drivers/power/sgm4154x_chg_mmi/Kconfig        | 23 ++++++
 drivers/power/sm5602_fg_mmi/Kconfig           |  3 +
 drivers/regulator/Kconfig                     |  4 +
 drivers/regulator/Makefile                    |  2 +
 drivers/regulator/wl2866d/Kconfig             |  9 +++
 drivers/regulator/wl2868c/Kconfig             |  9 +++
 drivers/usb/typec/Kconfig                     |  4 +
 drivers/usb/typec/Makefile                    |  2 +
 drivers/usb/typec/adapter_class/Kbuild        |  6 +-
 drivers/usb/typec/adapter_class/Kconfig       | 11 +++
 drivers/usb/typec/mmi_tcpc/Kbuild             | 20 ++---
 drivers/usb/typec/mmi_tcpc/Kconfig            | 50 ++++++++++++
 73 files changed, 604 insertions(+), 299 deletions(-)
 create mode 100644 drivers/backlight/Kconfig
 create mode 100644 drivers/backlight/Makefile
 create mode 100644 drivers/backlight/aw99703/Kconfig
 create mode 100644 drivers/backlight/lm3697/Kconfig
 create mode 100644 drivers/backlight/sm5350/Kconfig
 create mode 100644 drivers/fm/Kconfig
 create mode 100644 drivers/input/misc/focal_fps_mmi/Kconfig
 create mode 100644 drivers/input/misc/qpnp_power_on_mmi/Kconfig
 create mode 100644 drivers/input/misc/rbs_fps_mmi/Kconfig
 delete mode 100755 drivers/leds/Kbuild
 delete mode 100644 drivers/leds/trigger/Kbuild
 create mode 100644 drivers/misc/ldo_vibrator_mmi/Kconfig
 create mode 100644 drivers/nfc/samsung/Kbuild
 create mode 100644 drivers/power/bq2589x_chg_mmi/Kconfig
 create mode 100644 drivers/power/bq2597x_mmi/Kconfig
 create mode 100644 drivers/power/bq2597x_mmi_iio/Kconfig
 create mode 100644 drivers/power/cw2217b_fg_mmi/Kconfig
 create mode 100644 drivers/power/mmi-smbcharger-iio/Kconfig
 create mode 100644 drivers/power/mmi_discrete_charger/Kconfig
 create mode 100644 drivers/power/mmi_discrete_turbo_charger/Kconfig
 create mode 100644 drivers/power/mmi_parallel_charger_iio/Kconfig
 create mode 100644 drivers/power/mmi_sc8549/Kconfig
 create mode 100644 drivers/power/rt9426a_fg_mmi/Kconfig
 create mode 100644 drivers/power/sd77426_fg_mmi/Kconfig
 create mode 100644 drivers/power/sgm4154x_chg_mmi/Kconfig
 create mode 100644 drivers/power/sm5602_fg_mmi/Kconfig
 create mode 100644 drivers/regulator/wl2866d/Kconfig
 create mode 100644 drivers/regulator/wl2868c/Kconfig
 create mode 100644 drivers/usb/typec/adapter_class/Kconfig
 create mode 100644 drivers/usb/typec/mmi_tcpc/Kconfig

diff --git a/arch/arm64/Kconfig b/arch/arm64/Kconfig
index 78b6914be214..38b89a81882a 100644
--- a/arch/arm64/Kconfig
+++ b/arch/arm64/Kconfig
@@ -2021,3 +2021,9 @@ config MOT_SECURE_CAMERA
 config MOT_USB_DEBUG_LOG
 	bool "For usb debug log in factiry SW"
 	default n
+
+config MMI_QC3P_TURBO_CHARGER
+        bool "QC3P turbo charging"
+        default n
+        help
+        Enable it only for products that support QC3P charging.
diff --git a/drivers/Kconfig b/drivers/Kconfig
index d494e30841b0..6f676f4e4fd3 100644
--- a/drivers/Kconfig
+++ b/drivers/Kconfig
@@ -230,6 +230,8 @@ source "drivers/interconnect/Kconfig"
 
 source "drivers/counter/Kconfig"
 
+source "drivers/backlight/Kconfig"
+
 source "drivers/mmi_annotate/Kconfig"
 
 source "drivers/mmi_info/Kconfig"
@@ -246,4 +248,6 @@ source "drivers/watchdog_cpu_ctx/Kconfig"
 
 source "drivers/watchdogtest/Kconfig"
 
+source "drivers/fm/Kconfig"
+
 endmenu
diff --git a/drivers/Makefile b/drivers/Makefile
index 967f17c64e5c..b0a906c40a48 100644
--- a/drivers/Makefile
+++ b/drivers/Makefile
@@ -187,6 +187,7 @@ obj-$(CONFIG_GNSS)		+= gnss/
 obj-$(CONFIG_INTERCONNECT)	+= interconnect/
 obj-$(CONFIG_COUNTER)		+= counter/
 
+obj-y				+= backlight/
 obj-$(CONFIG_MMI_ANNOTATE)	+= mmi_annotate/
 obj-$(CONFIG_MMI_INFO)		+= mmi_info/
 obj-$(CONFIG_MMI_RELAY)	+= mmi_relay/
@@ -195,3 +196,4 @@ obj-$(CONFIG_SENSORS_CLASS)	+= sensors/
 obj-$(CONFIG_TZLOG_DUMP)	+= tzlog_dump/
 obj-$(CONFIG_WATCHDOG_CPU_CTX)	+= watchdog_cpu_ctx/
 obj-$(CONFIG_WATCHDOGTEST)	+= watchdogtest/
+obj-$(CONFIG_RADIO_ELNA)	+= fm/
diff --git a/drivers/backlight/Kconfig b/drivers/backlight/Kconfig
new file mode 100644
index 000000000000..cde255a30790
--- /dev/null
+++ b/drivers/backlight/Kconfig
@@ -0,0 +1,3 @@
+source "drivers/backlight/aw99703/Kconfig"
+source "drivers/backlight/lm3697/Kconfig"
+source "drivers/backlight/sm5350/Kconfig"
diff --git a/drivers/backlight/Makefile b/drivers/backlight/Makefile
new file mode 100644
index 000000000000..063a3080426b
--- /dev/null
+++ b/drivers/backlight/Makefile
@@ -0,0 +1,3 @@
+obj-y	+= aw99703/
+obj-y	+= lm3697/
+obj-y	+= sm5350/
diff --git a/drivers/backlight/aw99703/Kbuild b/drivers/backlight/aw99703/Kbuild
index 0772225e16dc..720ec9726b2f 100644
--- a/drivers/backlight/aw99703/Kbuild
+++ b/drivers/backlight/aw99703/Kbuild
@@ -2,4 +2,4 @@
 EXTRA_CFLAGS += -Wall
 EXTRA_CFLAGS += -I$(TOP)/motorola/kernel/modules/include
 
-obj-m += leds_aw99703.o
+obj-$(CONFIG_LEDS_AW99703) += leds_aw99703.o
diff --git a/drivers/backlight/aw99703/Kconfig b/drivers/backlight/aw99703/Kconfig
new file mode 100644
index 000000000000..720ed5c9c580
--- /dev/null
+++ b/drivers/backlight/aw99703/Kconfig
@@ -0,0 +1,9 @@
+config LEDS_AW99703
+        tristate "AW99703 LED driver"
+        default n
+        help
+         Say Y here if you have a aw99703 compatible
+         device connected.
+
+         This driver can also be built as a module.  If so, the module
+         will be called leds_aw99703.
diff --git a/drivers/backlight/lm3697/Kbuild b/drivers/backlight/lm3697/Kbuild
index 3204c539185a..5904da603dc7 100644
--- a/drivers/backlight/lm3697/Kbuild
+++ b/drivers/backlight/lm3697/Kbuild
@@ -2,7 +2,7 @@
 EXTRA_CFLAGS += -Wall
 EXTRA_CFLAGS += -I$(TOP)/motorola/kernel/modules/include
 
-obj-m := leds_lm3697.o
+obj-$(CONFIG_LEDS_LM3697_MMI) += leds_lm3697.o
 leds_lm3697-objs += ti_lm3697.o
 leds_lm3697-objs += ti_lm3697_backlight.o
 leds_lm3697-objs += ti_lm3697_backlight_data.o
diff --git a/drivers/backlight/lm3697/Kconfig b/drivers/backlight/lm3697/Kconfig
new file mode 100644
index 000000000000..9cbd1f0dd44c
--- /dev/null
+++ b/drivers/backlight/lm3697/Kconfig
@@ -0,0 +1,9 @@
+config LEDS_LM3697_MMI
+        tristate "LM3697 LED driver"
+        default n
+        help
+         Say Y here if you have a lm3697 compatible
+         device connected.
+
+         This driver can also be built as a module.  If so, the module
+         will be called leds_lm3697.
diff --git a/drivers/backlight/sm5350/Kbuild b/drivers/backlight/sm5350/Kbuild
index 41852f505da4..a1d981490792 100644
--- a/drivers/backlight/sm5350/Kbuild
+++ b/drivers/backlight/sm5350/Kbuild
@@ -1,5 +1,5 @@
 EXTRA_CFLAGS += -Wall
 EXTRA_CFLAGS += -I$(TOP)/motorola/kernel/modules/include
 
-obj-m   += sm5350_bl.o
+obj-$(CONFIG_LEDS_SM5350)   += sm5350_bl.o
 
diff --git a/drivers/backlight/sm5350/Kconfig b/drivers/backlight/sm5350/Kconfig
new file mode 100644
index 000000000000..8e141b2284c5
--- /dev/null
+++ b/drivers/backlight/sm5350/Kconfig
@@ -0,0 +1,6 @@
+config LEDS_SM5350
+	tristate "LED support for SM5350"
+	depends on LEDS_CLASS
+	depends on I2C
+	help
+	  This option enables support for LEDs connected to SM5350.
diff --git a/drivers/fm/Kbuild b/drivers/fm/Kbuild
index 3140ec437078..2a87f1669d32 100644
--- a/drivers/fm/Kbuild
+++ b/drivers/fm/Kbuild
@@ -1,4 +1 @@
-# add -Wall to try to catch everything we can.
-EXTRA_CFLAGS += -Wall
-
-obj-m += fm_ctrl.o
+obj-$(CONFIG_RADIO_ELNA) += fm_ctrl.o
diff --git a/drivers/fm/Kconfig b/drivers/fm/Kconfig
new file mode 100644
index 000000000000..0d4966aa36fe
--- /dev/null
+++ b/drivers/fm/Kconfig
@@ -0,0 +1,4 @@
+config RADIO_ELNA
+	tristate "ELNA FM Radio"
+	help
+	Choose Y here if you have this FM radio chip.
diff --git a/drivers/input/misc/Kconfig b/drivers/input/misc/Kconfig
index 7f39a7dece91..71d8b02806e0 100644
--- a/drivers/input/misc/Kconfig
+++ b/drivers/input/misc/Kconfig
@@ -914,4 +914,8 @@ config INPUT_STPMIC1_ONKEY
 	  To compile this driver as a module, choose M here: the
 	  module will be called stpmic1_onkey.
 
+source "drivers/input/misc/rbs_fps_mmi/Kconfig"
+source "drivers/input/misc/qpnp_power_on_mmi/Kconfig"
+source "drivers/input/misc/focal_fps_mmi/Kconfig"
+
 endif
diff --git a/drivers/input/misc/Makefile b/drivers/input/misc/Makefile
index b4e94422cea7..f5ccbe023814 100644
--- a/drivers/input/misc/Makefile
+++ b/drivers/input/misc/Makefile
@@ -88,6 +88,11 @@ obj-$(CONFIG_INPUT_XEN_KBDDEV_FRONTEND)	+= xen-kbdfront.o
 obj-$(CONFIG_INPUT_YEALINK)		+= yealink.o
 obj-$(CONFIG_INPUT_IDEAPAD_SLIDEBAR)	+= ideapad_slidebar.o
 obj-$(CONFIG_FINGERPRINT_GOODIX_FOD_MMI) += goodix_fod_mmi/
-obj-$(CONFIG_INPUT_AWINIC_HAPTIC) += awinic/
-obj-$(CONFIG_SARSENSOR_SX937X) += sx937x/
-obj-$(CONFIG_MOTO_UTAGS) += utag/
+obj-$(CONFIG_INPUT_CHIPONE_FPS_SPI)	+= chipone_fps_mmi/
+obj-$(CONFIG_INPUT_MISC_FPC1020)	+= fpc_fps_mmi/
+obj-$(CONFIG_INPUT_EGIS_FPS_MMI)	+= rbs_fps_mmi/
+obj-$(CONFIG_INPUT_SILEAD_FPS_MMI)	+= silead_fps_mmi/
+obj-$(CONFIG_INPUT_QPNP_POWER_ON_MMI)	+= qpnp_power_on_mmi/
+obj-$(CONFIG_INPUT_GOODIX_FOD_MMI)	+= goodix_fod_mmi/
+obj-$(CONFIG_INPUT_FOCAL_FPS_MMI)	+= focal_fps_mmi/
+obj-$(CONFIG_RBS_FOD_MMI)		+= rbs_fod_mmi/
diff --git a/drivers/input/misc/focal_fps_mmi/Kbuild b/drivers/input/misc/focal_fps_mmi/Kbuild
index 4f9cb1ae2a70..fc78fe8df6e3 100644
--- a/drivers/input/misc/focal_fps_mmi/Kbuild
+++ b/drivers/input/misc/focal_fps_mmi/Kbuild
@@ -1,49 +1,3 @@
-# add -Wall to try to catch everything we can.
-EXTRA_CFLAGS += -Wall
-EXTRA_CFLAGS += -I$(ANDROID_BUILD_TOP)/motorola/kernel/modules/include \
-		-I$(ANDROID_BUILD_TOP)/motorola/kernel/modules/drivers/input/misc/focal_fps_mmi
-
-
-
-
-
-
-
-# Early-logging level for ff_hal, ff_ta and driver as well. Note that this
-# logging level only takes effect before ff_config module is loaded.
-#
-# Options:
-# 0 - FF_LOG_LEVEL_ALL, All
-# 1 - FF_LOG_LEVEL_VBS, Verbose
-# 2 - FF_LOG_LEVEL_DBG, Debug
-# 3 - FF_LOG_LEVEL_INF, Info
-# 4 - FF_LOG_LEVEL_WRN, Warning
-# 5 - FF_LOG_LEVEL_ERR, Error
-EXTRA_CFLAGS  += -D__FF_BUILD_DATE=\"$(shell date +%Y%m%d)\"
-EXTRA_CFLAGS += -D__FF_EARLY_LOG_LEVEL=2
-
-
-focal_fps_mmi-y +=  ff_ctl.o plat-msm.o
-
-ifneq ($(filter m y,$(CONFIG_INPUT_EGISTEC_FPS_NAVI)),)
-	EXTRA_CFLAGS += -DCONFIG_INPUT_EGISTEC_FPS_NAVI
-	ets_fps_mmi-y += ets_navi_input.o
-endif
-ifneq ($(filter m y,$(CONFIG_INPUT_EGISTEC_FPS_NAVI_VERTICAL)),)
-	EXTRA_CFLAGS += -DCONFIG_INPUT_EGISTEC_FPS_NAVI_VERTICAL
-endif
-ifeq ($(TARGET_BUILD_VARIANT),userdebug)
-	EXTRA_CFLAGS += -DFP_SPI_DEBUG
-endif
-
-ifneq ($(filter m y,$(CONFIG_FOCAL_SUPPORT_DRM_SCREEN)),)
-	EXTRA_CFLAGS += -DCONFIG_DRM_PANEL_NOTIFICATIONS
-endif
-
-ifneq ($(filter m y, $(CONFIG_FOCAL_PANEL_NOTIFICATIONS)),)
-	EXTRA_CFLAGS += -DCONFIG_PANEL_NOTIFICATIONS
-endif
-
-obj-m += focal_fps_mmi.o
-
+focal_fps_mmi-y += ff_ctl.o plat-msm.o
 
+obj-$(CONFIG_INPUT_FOCAL_FPS_MMI) += focal_fps_mmi.o
diff --git a/drivers/input/misc/focal_fps_mmi/Kconfig b/drivers/input/misc/focal_fps_mmi/Kconfig
new file mode 100644
index 000000000000..049231c42957
--- /dev/null
+++ b/drivers/input/misc/focal_fps_mmi/Kconfig
@@ -0,0 +1,6 @@
+config INPUT_FOCAL_FPS_MMI
+	tristate "FOCAL fingerprint sensor MMI"
+	default n
+	help
+	 Say Y here if you wish to include support for MMI FOCAL finger print sensor.
+	 If unsure, say N.
diff --git a/drivers/input/misc/qpnp_power_on_mmi/Kbuild b/drivers/input/misc/qpnp_power_on_mmi/Kbuild
index dd859e89187d..61560eb32cfa 100644
--- a/drivers/input/misc/qpnp_power_on_mmi/Kbuild
+++ b/drivers/input/misc/qpnp_power_on_mmi/Kbuild
@@ -3,5 +3,5 @@ EXTRA_CFLAGS += -Wall
 EXTRA_CFLAGS += -I$(TOP)/motorola/kernel/modules/include
 KBUILD_EXTRA_SYMBOLS += $(CURDIR)/../../kernel/msm-$(MODULE_KERNEL_VERSION)/Module.symvers
 
-obj-m += qpnp-power-on-mmi.o
+obj-$(CONFIG_INPUT_QPNP_POWER_ON_MMI) += qpnp-power-on-mmi.o
 
diff --git a/drivers/input/misc/qpnp_power_on_mmi/Kconfig b/drivers/input/misc/qpnp_power_on_mmi/Kconfig
new file mode 100644
index 000000000000..0c1768d56cdc
--- /dev/null
+++ b/drivers/input/misc/qpnp_power_on_mmi/Kconfig
@@ -0,0 +1,8 @@
+config INPUT_QPNP_POWER_ON_MMI
+        tristate "QPNP PMIC Power-on MMI support"
+        depends on SPMI
+        help
+          This option enables device driver support for the power-on
+          functionality of Qualcomm Technologies, Inc. PNP PMICs.  It supports
+          reporting the change in status of the KPDPWR_N line (connected to the
+          power-key) as well as reset features.
diff --git a/drivers/input/misc/rbs_fps_mmi/Kbuild b/drivers/input/misc/rbs_fps_mmi/Kbuild
index b55d1b2190a6..a8e27d8688a9 100644
--- a/drivers/input/misc/rbs_fps_mmi/Kbuild
+++ b/drivers/input/misc/rbs_fps_mmi/Kbuild
@@ -1,18 +1,12 @@
-# add -Wall to try to catch everything we can.
-EXTRA_CFLAGS += -Wall
-EXTRA_CFLAGS += -I$(ANDROID_BUILD_TOP)/motorola/kernel/modules/include \
-				-I$(ANDROID_BUILD_TOP)/motorola/kernel/modules/drivers/input/misc/rbs_fps_mmi
-
 ifneq ($(filter m y,$(CONFIG_EGIS_PANEL_NOTIFICATIONS)),)
-    EXTRA_CFLAGS += -DCONFIG_PANEL_NOTIFICATIONS
+    ccflags-y += -DCONFIG_PANEL_NOTIFICATIONS
 endif
 
 ifneq ($(filter m y,$(CONFIG_ET721_FOD)),)
-    EXTRA_CFLAGS += -DET721_FOD
+    ccflags-y += -DET721_FOD
 endif
 
-rbs_fps_mmi-y += etxxx_fp.o
-
+rbs_fps_mmi-objs += etxxx_fp.o
 
-obj-m += rbs_fps_mmi.o
+obj-$(CONFIG_INPUT_EGIS_FPS_MMI) += rbs_fps_mmi.o
 
diff --git a/drivers/input/misc/rbs_fps_mmi/Kconfig b/drivers/input/misc/rbs_fps_mmi/Kconfig
new file mode 100644
index 000000000000..5944efc4d758
--- /dev/null
+++ b/drivers/input/misc/rbs_fps_mmi/Kconfig
@@ -0,0 +1,25 @@
+config INPUT_EGIS_FPS_MMI
+	tristate "EGIS fingerprint sensor MMI"
+	default n
+	help
+	 Say Y here if you wish to include support for MMI EGIS finger print sensor.
+	 If unsure, say N.
+
+config ET721_FOD
+	bool "FOD Support for egis"
+	default n
+	depends on INPUT_EGIS_FPS_MMI
+	help
+	 Say Y here if you wish to include support for MMI EGIS finger print sensor
+         FOD support.
+	 If unsure, say N.
+
+config EGIS_PANEL_NOTIFICATIONS
+	bool "Panel Notification Support for egis"
+	default n
+	depends on INPUT_EGIS_FPS_MMI
+	help
+	 Say Y here if you wish to include support for MMI EGIS finger print sensor
+         panel notification framework support.
+	 If unsure, say N.
+
diff --git a/drivers/input/touchscreen/Kconfig b/drivers/input/touchscreen/Kconfig
index f8dbc3e868e7..fd4668871b3e 100644
--- a/drivers/input/touchscreen/Kconfig
+++ b/drivers/input/touchscreen/Kconfig
@@ -1337,4 +1337,6 @@ source "drivers/input/touchscreen/stmicro_mmi/Kconfig"
 
 source "drivers/input/touchscreen/touchscreen_mmi/Kconfig"
 
+source "drivers/input/touchscreen/goodix_berlin_mmi/Kconfig"
+
 endif
diff --git a/drivers/input/touchscreen/Makefile b/drivers/input/touchscreen/Makefile
index dfe8da830025..9c90e416d043 100644
--- a/drivers/input/touchscreen/Makefile
+++ b/drivers/input/touchscreen/Makefile
@@ -117,3 +117,4 @@ obj-$(CONFIG_TOUCHSCREEN_SYNAPTICS_DSX) += synaptics_dsx/
 obj-$(CONFIG_TOUCHSCREEN_NT36XXX)	+= nt36xxx/
 obj-$(CONFIG_INPUT_TOUCHSCREEN_MMI)	+= touchscreen_mmi/
 obj-$(CONFIG_TOUCHSCREEN_FTS_TS)	+= stmicro_mmi/
+obj-$(CONFIG_TOUCHSCREEN_GOODIX_BRL)	+= goodix_berlin_mmi/
diff --git a/drivers/input/touchscreen/goodix_berlin_mmi/Kbuild b/drivers/input/touchscreen/goodix_berlin_mmi/Kbuild
index 65bf4824597e..cf2c219897c3 100644
--- a/drivers/input/touchscreen/goodix_berlin_mmi/Kbuild
+++ b/drivers/input/touchscreen/goodix_berlin_mmi/Kbuild
@@ -3,7 +3,7 @@ EXTRA_CFLAGS += -Wall
 EXTRA_CFLAGS += \
 		-I$(ANDROID_BUILD_TOP)/motorola/kernel/modules/include
 
-obj-m := goodix_brl_mmi.o
+obj-$(CONFIG_TOUCHSCREEN_GOODIX_BRL) := goodix_brl_mmi.o
 goodix_brl_mmi-objs := \
 					goodix_brl_i2c.o \
 					goodix_brl_spi.o \
@@ -17,117 +17,77 @@ goodix_brl_mmi-objs := \
 					goodix_ts_config.o
 
 ifneq ($(filter m y, $(CONFIG_DRM_PANEL_EVENT_NOTIFICATIONS)),)
-	EXTRA_CFLAGS += -DCONFIG_DRM_PANEL_EVENT_NOTIFICATIONS
+	ccflags-y += -DCONFIG_DRM_PANEL_EVENT_NOTIFICATIONS
 endif
 
 ifneq ($(filter m y,$(CONFIG_DRM_PANEL_NOTIFICATIONS)),)
-    EXTRA_CFLAGS += -DCONFIG_DRM_PANEL_NOTIFICATIONS
+    ccflags-y += -DCONFIG_DRM_PANEL_NOTIFICATIONS
 endif
 
 ifneq ($(filter m y,$(CONFIG_TOUCHSCREEN_GOODIX_BRL_SPI)),)
-    EXTRA_CFLAGS += -DCONFIG_TOUCHSCREEN_GOODIX_BRL_SPI
+    ccflags-y += -DCONFIG_TOUCHSCREEN_GOODIX_BRL_SPI
 endif
 
 ifneq ($(filter m y,$(CONFIG_GTP_FOD)),)
-    EXTRA_CFLAGS += -DCONFIG_GTP_FOD
+    ccflags-y += -DCONFIG_GTP_FOD
     goodix_brl_mmi-objs += goodix_ts_gesture_fod.o
 else
     goodix_brl_mmi-objs += goodix_ts_gesture.o
 endif
 
 ifneq ($(filter m y,$(CONFIG_GTP_LIMIT_USE_SUPPLIER)),)
-    EXTRA_CFLAGS += -DCONFIG_GTP_LIMIT_USE_SUPPLIER
+    ccflags-y += -DCONFIG_GTP_LIMIT_USE_SUPPLIER
 endif
 
 ifneq ($(filter m y,$(CONFIG_GTP_ENABLE_PM_QOS)),)
-    EXTRA_CFLAGS += -DCONFIG_GTP_ENABLE_PM_QOS
+    ccflags-y += -DCONFIG_GTP_ENABLE_PM_QOS
 endif
 
 ifneq ($(filter m y,$(CONFIG_GOODIX_ESD_ENABLE)),)
-    EXTRA_CFLAGS += -DCONFIG_GOODIX_ESD_ENABLE
+    ccflags-y += -DCONFIG_GOODIX_ESD_ENABLE
 endif
 
 ifneq ($(filter m y,$(CONFIG_TARGET_BUILD_FACROTY)),)
-    EXTRA_CFLAGS += -DGTP_SAVE_IN_CSV
+    ccflags-y += -DGTP_SAVE_IN_CSV
 endif
 
 ifneq ($(filter m y,$(CONFIG_GTP_LAST_TIME)),)
-    EXTRA_CFLAGS += -DCONFIG_GTP_LAST_TIME
+    ccflags-y += -DCONFIG_GTP_LAST_TIME
 endif
 
 ifneq ($(filter m y,$(CONFIG_GTP_DDA_STYLUS)),)
-    EXTRA_CFLAGS += -DCONFIG_GTP_DDA_STYLUS
+    ccflags-y += -DCONFIG_GTP_DDA_STYLUS
     goodix_brl_mmi-objs += goodix_ts_dda.o
 endif
 
 ifneq ($(filter m y,$(CONFIG_MOTO_DDA_PASSIVE_STYLUS)),)
-    EXTRA_CFLAGS += -DCONFIG_MOTO_DDA_PASSIVESTYLUS
+    ccflags-y += -DCONFIG_MOTO_DDA_PASSIVESTYLUS
 endif
 
 ifneq ($(filter m y,$(CONFIG_INPUT_TOUCHSCREEN_MMI)),)
-    EXTRA_CFLAGS += -DCONFIG_INPUT_TOUCHSCREEN_MMI
+    ccflags-y += -DCONFIG_INPUT_TOUCHSCREEN_MMI
     goodix_brl_mmi-objs += goodix_ts_mmi.o
 endif
 
 ifneq ($(filter m y, $(CONFIG_BOARD_USES_DOUBLE_TAP_CTRL)),)
-	EXTRA_CFLAGS += -DCONFIG_BOARD_USES_DOUBLE_TAP_CTRL
+	ccflags-y += -DCONFIG_BOARD_USES_DOUBLE_TAP_CTRL
 endif
 
 ifneq ($(filter m y,$(CONFIG_INPUT_GOODIX_MMI_ENABLE_PALM)),)
-    EXTRA_CFLAGS += -DGOODIX_PALM_SENSOR_EN
+    ccflags-y += -DGOODIX_PALM_SENSOR_EN
 endif
 
 ifeq ($(TARGET_BUILD_VARIANT), userdebug)
 ifneq ($(filter m y, $(CONFIG_GTP_GHOST_LOG_CAPTURE)),)
-	EXTRA_CFLAGS += -DCONFIG_GTP_GHOST_LOG_CAPTURE
+	ccflags-y += -DCONFIG_GTP_GHOST_LOG_CAPTURE
 	goodix_brl_mmi-objs += goodix_ts_log_capture.o
 endif
 endif
 
 ifneq ($(filter m y,$(CONFIG_ENABLE_GTP_PALM_CANCEL)),)
-    EXTRA_CFLAGS += -DCONFIG_ENABLE_GTP_PALM_CANCEL
+    ccflags-y += -DCONFIG_ENABLE_GTP_PALM_CANCEL
 endif
 
 ifneq ($(filter m y,$(CONFIG_INPUT_GTP_PASSIVE_PEN_NOTIFIER)),)
-	EXTRA_CFLAGS += -DGTP_PEN_NOTIFIER
+	ccflags-y += -DGTP_PEN_NOTIFIER
 endif
-
-ifneq ($(findstring _factory,$(TARGET_PRODUCT)),)
-USE_TARGET_PRODUCT := $(strip $(subst _factory,, $(TARGET_PRODUCT)))
-else
-USE_TARGET_PRODUCT := $(TARGET_PRODUCT)
-endif
-
-ifneq ($(filter oneli, $(USE_TARGET_PRODUCT)), )
-        EXTRA_CFLAGS += -DPRODUCT_ONELI
-else ifneq ($(filter hiphi, $(USE_TARGET_PRODUCT)), )
-        EXTRA_CFLAGS += -DPRODUCT_HIPHI
-else ifneq ($(filter hiphic, $(USE_TARGET_PRODUCT)), )
-        EXTRA_CFLAGS += -DPRODUCT_HIPHIC
-else ifneq ($(filter li, $(USE_TARGET_PRODUCT)), )
-        EXTRA_CFLAGS += -DPRODUCT_LI
-else ifneq ($(filter eqs, $(USE_TARGET_PRODUCT)), )
-        EXTRA_CFLAGS += -DPRODUCT_EQS
-else ifneq ($(filter tundra%, $(USE_TARGET_PRODUCT)), )
-        EXTRA_CFLAGS += -DPRODUCT_TUNDRA
-else ifneq ($(filter miami%, $(USE_TARGET_PRODUCT)), )
-        EXTRA_CFLAGS += -DPRODUCT_MIAMI
-else ifneq ($(filter rtwo%, $(USE_TARGET_PRODUCT)), )
-        EXTRA_CFLAGS += -DPRODUCT_RTWO
-else ifneq ($(filter ctwo%, $(USE_TARGET_PRODUCT)), )
-        EXTRA_CFLAGS += -DPRODUCT_CTWO
-else ifneq ($(filter bronco, $(USE_TARGET_PRODUCT)), )
-        EXTRA_CFLAGS += -DPRODUCT_BRONCO
-else ifneq ($(filter felix, $(USE_TARGET_PRODUCT)), )
-        EXTRA_CFLAGS += -DCONFIG_GTP_DISP_MODE
-else ifneq ($(filter zeekr%, $(USE_TARGET_PRODUCT)), )
-        EXTRA_CFLAGS += -DPRODUCT_ZEEKR
-else ifneq ($(filter lynkco%, $(USE_TARGET_PRODUCT)), )
-        EXTRA_CFLAGS += -DPRODUCT_LYNKCO
-else ifneq ($(filter bangkk%, $(USE_TARGET_PRODUCT)), )
-        EXTRA_CFLAGS += -DPRODUCT_BANGKK
-endif
-$(warning "william EXTRA_CFLAGS = $(EXTRA_CFLAGS)")
-
-KBUILD_EXTRA_SYMBOLS += $(CURDIR)/$(KBUILD_EXTMOD)/../touchscreen_mmi/$(GKI_OBJ_MODULE_DIR)/Module.symvers
-KBUILD_EXTRA_SYMBOLS += $(CURDIR)/$(KBUILD_EXTMOD)/../../../mmi_info/$(GKI_OBJ_MODULE_DIR)/Module.symvers
diff --git a/drivers/input/touchscreen/goodix_berlin_mmi/Kconfig b/drivers/input/touchscreen/goodix_berlin_mmi/Kconfig
index 223d5aea824e..6e512e5e3c3f 100644
--- a/drivers/input/touchscreen/goodix_berlin_mmi/Kconfig
+++ b/drivers/input/touchscreen/goodix_berlin_mmi/Kconfig
@@ -17,4 +17,23 @@ config TOUCHSCREEN_GOODIX_BRL_SPI
 	help
 		Say Y here if the touchscreen is connected via SPI bus.
 
+config TOUCHSCREEN_GOODIX_BRL_I2C
+	tristate "Goodix berlin touchscreen"
+	help
+	  Say Y here if you have a Goodix berlin series touch controller
+	  to your system.
+
+	  If build module, say M.
+	  If unsure, say N.
+
+config GTP_FOD
+	bool "supports FOD"
+	help
+		Say Y here to support FOD.
+
+config GTP_LIMIT_USE_SUPPLIER
+	bool "supports using supplier"
+	help
+		Say Y here to support using supplier.
+
 endif
diff --git a/drivers/leds/Kbuild b/drivers/leds/Kbuild
deleted file mode 100755
index eb1a7d0ae20c..000000000000
--- a/drivers/leds/Kbuild
+++ /dev/null
@@ -1,6 +0,0 @@
-# add -Wall to try to catch everything we can.
-EXTRA_CFLAGS += -Wall
-EXTRA_CFLAGS += -I$(TOP)/motorola/kernel/modules/include
-
-
-obj-m += leds-indicator-pwm.o
diff --git a/drivers/leds/trigger/Kbuild b/drivers/leds/trigger/Kbuild
deleted file mode 100644
index 1983f4f1e905..000000000000
--- a/drivers/leds/trigger/Kbuild
+++ /dev/null
@@ -1,5 +0,0 @@
-EXTRA_CFLAGS += -Wall
-EXTRA_CFLAGS += -I$(TOP)/motorola/kernel/modules/include
-
-
-obj-m += ledtrig-sol.o
diff --git a/drivers/leds/trigger/Kconfig b/drivers/leds/trigger/Kconfig
index ce9429ca6dde..0a80142fbae1 100644
--- a/drivers/leds/trigger/Kconfig
+++ b/drivers/leds/trigger/Kconfig
@@ -144,4 +144,11 @@ config LEDS_TRIGGER_AUDIO
 	  the audio mute and mic-mute changes.
 	  If unsure, say N
 
+config LEDS_TRIGGER_SOL
+	tristate "Sign of Life LED Trigger"
+	help
+	  Configurable LED trigger that alters the LED blink
+	  pattern based on power supply status.
+	  If unsure, say N
+
 endif # LEDS_TRIGGERS
diff --git a/drivers/leds/trigger/Makefile b/drivers/leds/trigger/Makefile
index a087dacd15ad..efc874937ebb 100644
--- a/drivers/leds/trigger/Makefile
+++ b/drivers/leds/trigger/Makefile
@@ -1,10 +1,18 @@
-KERNEL_SRC ?= /lib/modules/$(shell uname -r)/build
-
-all:
-	$(MAKE) -C $(KERNEL_SRC) M=$(M) modules $(KBUILD_OPTIONS)
-
-modules_install:
-	$(MAKE) INSTALL_MOD_STRIP=1 -C $(KERNEL_SRC) M=$(M) modules_install
-
-clean:
-	$(MAKE) -C $(KERNEL_SRC) M=$(PWD) clean
+# SPDX-License-Identifier: GPL-2.0
+obj-$(CONFIG_LEDS_TRIGGER_TIMER)	+= ledtrig-timer.o
+obj-$(CONFIG_LEDS_TRIGGER_ONESHOT)	+= ledtrig-oneshot.o
+obj-$(CONFIG_LEDS_TRIGGER_DISK)		+= ledtrig-disk.o
+obj-$(CONFIG_LEDS_TRIGGER_MTD)		+= ledtrig-mtd.o
+obj-$(CONFIG_LEDS_TRIGGER_HEARTBEAT)	+= ledtrig-heartbeat.o
+obj-$(CONFIG_LEDS_TRIGGER_BACKLIGHT)	+= ledtrig-backlight.o
+obj-$(CONFIG_LEDS_TRIGGER_GPIO)		+= ledtrig-gpio.o
+obj-$(CONFIG_LEDS_TRIGGER_CPU)		+= ledtrig-cpu.o
+obj-$(CONFIG_LEDS_TRIGGER_ACTIVITY)	+= ledtrig-activity.o
+obj-$(CONFIG_LEDS_TRIGGER_DEFAULT_ON)	+= ledtrig-default-on.o
+obj-$(CONFIG_LEDS_TRIGGER_TRANSIENT)	+= ledtrig-transient.o
+obj-$(CONFIG_LEDS_TRIGGER_CAMERA)	+= ledtrig-camera.o
+obj-$(CONFIG_LEDS_TRIGGER_PANIC)	+= ledtrig-panic.o
+obj-$(CONFIG_LEDS_TRIGGER_NETDEV)	+= ledtrig-netdev.o
+obj-$(CONFIG_LEDS_TRIGGER_PATTERN)	+= ledtrig-pattern.o
+obj-$(CONFIG_LEDS_TRIGGER_AUDIO)	+= ledtrig-audio.o
+obj-$(CONFIG_LEDS_TRIGGER_SOL)		+= ledtrig-sol.o
diff --git a/drivers/misc/Kconfig b/drivers/misc/Kconfig
index 9ba14d11dbe3..965cef9324f3 100644
--- a/drivers/misc/Kconfig
+++ b/drivers/misc/Kconfig
@@ -543,4 +543,5 @@ source "drivers/misc/awinic/Kconfig"
 source "drivers/misc/mmi_sys_temp/Kconfig"
 source "drivers/misc/sx937x/Kconfig"
 source "drivers/misc/utag/Kconfig"
+source "drivers/misc/ldo_vibrator_mmi/Kconfig"
 endmenu
diff --git a/drivers/misc/Makefile b/drivers/misc/Makefile
index 34b09dd815a7..288d85c1f0c8 100644
--- a/drivers/misc/Makefile
+++ b/drivers/misc/Makefile
@@ -1,10 +1,72 @@
-KERNEL_SRC ?= /lib/modules/$(shell uname -r)/build
+# SPDX-License-Identifier: GPL-2.0
+#
+# Makefile for misc devices that really don't fit anywhere else.
+#
 
-all:
-	$(MAKE) -C $(KERNEL_SRC) M=$(shell pwd) modules $(KBUILD_OPTIONS)
-
-modules_install:
-	$(MAKE) INSTALL_MOD_STRIP=1 -C $(KERNEL_SRC) M=$(shell pwd) modules_install
-
-clean:
-	$(MAKE) -C $(KERNEL_SRC) M=$(PWD) clean
\ No newline at end of file
+obj-$(CONFIG_IBM_ASM)		+= ibmasm/
+obj-$(CONFIG_IBMVMC)		+= ibmvmc.o
+obj-$(CONFIG_AD525X_DPOT)	+= ad525x_dpot.o
+obj-$(CONFIG_AD525X_DPOT_I2C)	+= ad525x_dpot-i2c.o
+obj-$(CONFIG_AD525X_DPOT_SPI)	+= ad525x_dpot-spi.o
+obj-$(CONFIG_INTEL_MID_PTI)	+= pti.o
+obj-$(CONFIG_ATMEL_SSC)		+= atmel-ssc.o
+obj-$(CONFIG_ATMEL_TCLIB)	+= atmel_tclib.o
+obj-$(CONFIG_DUMMY_IRQ)		+= dummy-irq.o
+obj-$(CONFIG_ICS932S401)	+= ics932s401.o
+obj-$(CONFIG_LKDTM)		+= lkdtm/
+obj-$(CONFIG_TIFM_CORE)       	+= tifm_core.o
+obj-$(CONFIG_TIFM_7XX1)       	+= tifm_7xx1.o
+obj-$(CONFIG_PHANTOM)		+= phantom.o
+obj-$(CONFIG_QCOM_COINCELL)	+= qcom-coincell.o
+obj-$(CONFIG_QCOM_FASTRPC)	+= fastrpc.o
+obj-$(CONFIG_SENSORS_BH1770)	+= bh1770glc.o
+obj-$(CONFIG_SENSORS_APDS990X)	+= apds990x.o
+obj-$(CONFIG_ENCLOSURE_SERVICES) += enclosure.o
+obj-$(CONFIG_KGDB_TESTS)	+= kgdbts.o
+obj-$(CONFIG_SGI_XP)		+= sgi-xp/
+obj-$(CONFIG_SGI_GRU)		+= sgi-gru/
+obj-$(CONFIG_CS5535_MFGPT)	+= cs5535-mfgpt.o
+obj-$(CONFIG_HP_ILO)		+= hpilo.o
+obj-$(CONFIG_APDS9802ALS)	+= apds9802als.o
+obj-$(CONFIG_ISL29003)		+= isl29003.o
+obj-$(CONFIG_ISL29020)		+= isl29020.o
+obj-$(CONFIG_SENSORS_TSL2550)	+= tsl2550.o
+obj-$(CONFIG_DS1682)		+= ds1682.o
+obj-$(CONFIG_C2PORT)		+= c2port/
+obj-$(CONFIG_HMC6352)		+= hmc6352.o
+obj-y				+= eeprom/
+obj-y				+= cb710/
+obj-$(CONFIG_VMWARE_BALLOON)	+= vmw_balloon.o
+obj-$(CONFIG_PCH_PHUB)		+= pch_phub.o
+obj-y				+= ti-st/
+obj-y				+= lis3lv02d/
+obj-$(CONFIG_ALTERA_STAPL)	+=altera-stapl/
+obj-$(CONFIG_INTEL_MEI)		+= mei/
+obj-$(CONFIG_VMWARE_VMCI)	+= vmw_vmci/
+obj-$(CONFIG_LATTICE_ECP3_CONFIG)	+= lattice-ecp3-config.o
+obj-$(CONFIG_SRAM)		+= sram.o
+obj-$(CONFIG_SRAM_EXEC)		+= sram-exec.o
+obj-y				+= mic/
+obj-$(CONFIG_GENWQE)		+= genwqe/
+obj-$(CONFIG_ECHO)		+= echo/
+obj-$(CONFIG_HDCP_QSEECOM)	+= hdcp_qseecom.o
+obj-$(CONFIG_VEXPRESS_SYSCFG)	+= vexpress-syscfg.o
+obj-$(CONFIG_CXL_BASE)		+= cxl/
+obj-$(CONFIG_PCI_ENDPOINT_TEST)	+= pci_endpoint_test.o
+obj-$(CONFIG_OCXL)		+= ocxl/
+obj-y				+= cardreader/
+obj-$(CONFIG_PVPANIC)   	+= pvpanic.o
+obj-$(CONFIG_HABANA_AI)		+= habanalabs/
+obj-$(CONFIG_XILINX_SDFEC)	+= xilinx_sdfec.o
+obj-$(CONFIG_UID_SYS_STATS)	+= uid_sys_stats.o
+obj-$(CONFIG_QSEECOM)		+= qseecom-mod.o
+qseecom-mod-y			:= qseecom.o
+qseecom-mod-$(CONFIG_COMPAT)	+= compat_qseecom.o
+obj-$(CONFIG_PROFILER)			+= profiler.o
+obj-$(CONFIG_WIGIG_SENSING_SPI)	+= wigig_sensing.o
+obj-$(CONFIG_HISI_HIKEY_USB)	+= hisi_hikey_usb.o
+obj-y				+= pen/
+obj-$(CONFIG_MOTO_UTAGS) 	+= utag/
+obj-$(CONFIG_MMI_SYS_TEMP)	+= mmi_sys_temp/
+obj-$(CONFIG_LDO_VIBRATOR_MMI)	+= ldo_vibrator_mmi/
+obj-$(CONFIG_SARSENSOR_SX937X)		+= sx937x/
diff --git a/drivers/misc/awinic/Kconfig b/drivers/misc/awinic/Kconfig
index dcb9e22a5ef1..7bc110c62b11 100644
--- a/drivers/misc/awinic/Kconfig
+++ b/drivers/misc/awinic/Kconfig
@@ -8,4 +8,4 @@ menuconfig INPUT_AWINIC_HAPTIC
 	help
 	  Enables Awinic Haptic driver.
 
-source "drivers/misc/awinic/aw8695/Kconfig"
\ No newline at end of file
+source "drivers/misc/awinic/aw8695/Kconfig"
diff --git a/drivers/misc/ldo_vibrator_mmi/Kconfig b/drivers/misc/ldo_vibrator_mmi/Kconfig
new file mode 100644
index 000000000000..912d94fb2267
--- /dev/null
+++ b/drivers/misc/ldo_vibrator_mmi/Kconfig
@@ -0,0 +1,3 @@
+config LDO_VIBRATOR_MMI
+	tristate "ldo vibrator driver"
+	default n
diff --git a/drivers/mmi_annotate/Kbuild b/drivers/mmi_annotate/Kbuild
index f554a4e253b2..76a40f6d5ecc 100644
--- a/drivers/mmi_annotate/Kbuild
+++ b/drivers/mmi_annotate/Kbuild
@@ -1,5 +1 @@
-# add -Wall to try to catch everything we can.
-EXTRA_CFLAGS += -Wall
-EXTRA_CFLAGS += -I$(ANDROID_BUILD_TOP)/motorola/kernel/modules/include
-
-obj-m += mmi_annotate.o
\ No newline at end of file
+obj-$(CONFIG_MMI_ANNOTATE) += mmi_annotate.o
diff --git a/drivers/mmi_info/Kbuild b/drivers/mmi_info/Kbuild
index 4a3f1a0f3fac..7c2d0eba268d 100644
--- a/drivers/mmi_info/Kbuild
+++ b/drivers/mmi_info/Kbuild
@@ -1,12 +1,6 @@
-# add -Wall to try to catch everything we can.
-EXTRA_CFLAGS += -Wall
-EXTRA_CFLAGS += -I$(ANDROID_BUILD_TOP)/motorola/kernel/modules/include
-
 mmi_info-objs += mmi_storage_info.o
 mmi_info-objs += mmi_ram_info.o
 mmi_info-objs += mmi_unit_info.o
 mmi_info-objs += mmi_boot_info.o
 mmi_info-objs += mmi_info_main.o
-obj-m += mmi_info.o
-
-KBUILD_EXTRA_SYMBOLS += $(CURDIR)/$(KBUILD_EXTMOD)/../mmi_annotate/$(GKI_OBJ_MODULE_DIR)/Module.symvers
+obj-$(CONFIG_MMI_INFO) += mmi_info.o
diff --git a/drivers/mmi_relay/Kbuild b/drivers/mmi_relay/Kbuild
index 12e00ae16d00..aaa4e173f7fb 100644
--- a/drivers/mmi_relay/Kbuild
+++ b/drivers/mmi_relay/Kbuild
@@ -1,5 +1 @@
-# add -Wall to try to catch everything we can.
-EXTRA_CFLAGS += -Wall
-EXTRA_CFLAGS += -I$(ANDROID_BUILD_TOP)/motorola/kernel/modules/include
-
-obj-m += mmi_relay.o
+obj-$(CONFIG_MMI_RELAY) += mmi_relay.o
diff --git a/drivers/moto_f_usbnet/Kbuild b/drivers/moto_f_usbnet/Kbuild
index 331c29b3cfa2..a016a4a552df 100644
--- a/drivers/moto_f_usbnet/Kbuild
+++ b/drivers/moto_f_usbnet/Kbuild
@@ -3,4 +3,4 @@ EXTRA_CFLAGS += -Wall
 EXTRA_CFLAGS += -I$(ANDROID_BUILD_TOP)/motorola/kernel/modules/include
 
 moto_f_usbnet-objs += f_usbnet.o
-obj-m += moto_f_usbnet.o
+obj-$(CONFIG_MOTO_F_USBNET) += moto_f_usbnet.o
diff --git a/drivers/nfc/samsung/Kbuild b/drivers/nfc/samsung/Kbuild
new file mode 100644
index 000000000000..56c1e5cb017d
--- /dev/null
+++ b/drivers/nfc/samsung/Kbuild
@@ -0,0 +1,8 @@
+ifneq ($(filter m y,$(CONFIG_SEC_NFC_PRODUCT_N5)),)
+	ccflags-y += -DCONFIG_SEC_NFC_PRODUCT_N5
+endif
+ifneq ($(filter m y,$(CONFIG_SEC_NFC_IF_I2C)),)
+	ccflags-y += -DCONFIG_SEC_NFC_IF_I2C
+endif
+
+obj-$(CONFIG_SEC_NFC) += sec_nfc.o
diff --git a/drivers/power/Kconfig b/drivers/power/Kconfig
index af8b08660ce2..111a8e5f4da0 100644
--- a/drivers/power/Kconfig
+++ b/drivers/power/Kconfig
@@ -7,3 +7,13 @@ source "drivers/power/bm_adsp_ulog/Kconfig"
 source "drivers/power/mmi_charger/Kconfig"
 source "drivers/power/qpnp_adaptive_charge/Kconfig"
 
+source "drivers/power/bq2597x_mmi_iio/Kconfig"
+source "drivers/power/bq2589x_chg_mmi/Kconfig"
+source "drivers/power/cw2217b_fg_mmi/Kconfig"
+source "drivers/power/sgm4154x_chg_mmi/Kconfig"
+source "drivers/power/sm5602_fg_mmi/Kconfig"
+source "drivers/power/mmi_discrete_charger/Kconfig"
+source "drivers/power/mmi_discrete_turbo_charger/Kconfig"
+source "drivers/power/mmi_parallel_charger_iio/Kconfig"
+source "drivers/power/mmi_sc8549/Kconfig"
+source "drivers/power/mmi-smbcharger-iio/Kconfig"
diff --git a/drivers/power/Makefile b/drivers/power/Makefile
index 75377b6d1d5c..53da74d36b85 100644
--- a/drivers/power/Makefile
+++ b/drivers/power/Makefile
@@ -3,6 +3,19 @@ obj-$(CONFIG_POWER_AVS)		+= avs/
 obj-$(CONFIG_POWER_RESET)	+= reset/
 obj-$(CONFIG_POWER_SUPPLY)	+= supply/
 
-obj-$(CONFIG_UBM_ADSP_ULOG	+= bm_adsp_ulog/
-obj-$(CONFIG_USE_MMI_CHARGER)	+= mmi_charger/
+obj-$(CONFIG_UBM_ADSP_ULOG)	+= bm_adsp_ulog/
+obj-$(CONFIG_MMI_CHARGER)	+= mmi_charger/
 obj-$(CONFIG_QPNP_ADAPTIVE_CHARGE)	+= qpnp_adaptive_charge/
+obj-$(CONFIG_MMI_PARALLEL_CHARGER_IIO) += mmi_parallel_charger_iio/
+obj-$(CONFIG_MMI_SC8549)	+= mmi_sc8549/
+obj-$(CONFIG_MMI_SMBCHARGER_IIO) += mmi-smbcharger-iio/
+obj-$(CONFIG_QPNP_ADAPTIVE_CHARGE) += qpnp_adaptive_charge/
+obj-$(CONFIG_BQ2597X_MMI) += bq2597x_mmi/
+obj-$(CONFIG_BQ2597X_MMI_IIO) += bq2597x_mmi_iio/
+obj-$(CONFIG_BQ2589X_CHARGER)	+= bq2589x_chg_mmi/
+obj-$(CONFIG_CW2217B_FG_MMI)	+= cw2217b_fg_mmi/
+obj-$(CONFIG_SGM4154X_CHARGER)	+= sgm4154x_chg_mmi/
+obj-$(CONFIG_MMI_SM5602_FG)	+= sm5602_fg_mmi/
+obj-$(CONFIG_MMI_DISCRETE_CHARGER) += mmi_discrete_charger/
+obj-$(CONFIG_MMI_DISCRETE_TURBO_CHARGER) += mmi_discrete_turbo_charger/
+obj-$(CONFIG_MMI_MM8013C_FG)	+= mm8013c_fg_mmi/
\ No newline at end of file
diff --git a/drivers/power/bq2589x_chg_mmi/Kbuild b/drivers/power/bq2589x_chg_mmi/Kbuild
index d40823873eee..e0e0802affc1 100644
--- a/drivers/power/bq2589x_chg_mmi/Kbuild
+++ b/drivers/power/bq2589x_chg_mmi/Kbuild
@@ -1,20 +1 @@
-# add -Wall to try to catch everything we can.
-EXTRA_CFLAGS += -Wall
-EXTRA_CFLAGS += -I$(ANDROID_BUILD_TOP)/motorola/kernel/modules/include \
-		-I$(ANDROID_BUILD_TOP)/motorola/kernel/modules/drivers/power/bq2589x_chg_mmi
-
-obj-m += bq2589x_charger.o
-
-ifneq ($(filter m y,$(CONFIG_MMI_QC3P_TURBO_CHARGER)),)
-        EXTRA_CFLAGS += -DCONFIG_MMI_QC3P_TURBO_CHARGER
-endif
-
-ifneq ($(filter m y,$(CONFIG_MMI_QC3P_WT6670_DETECTED)),)
-        EXTRA_CFLAGS += -DCONFIG_MMI_QC3P_WT6670_DETECTED
-endif
-
-ifneq ($(filter m y,$(CONFIG_MMI_EXT_CHG_LED)),)
-        EXTRA_CFLAGS += -DCONFIG_MMI_EXT_CHG_LED
-endif
-
-KBUILD_EXTRA_SYMBOLS += $(CURDIR)/$(KBUILD_EXTMOD)/../mmi_discrete_charger/$(GKI_OBJ_MODULE_DIR)/Module.symvers
+obj-$(CONFIG_BQ2589X_CHARGER) += bq2589x_charger.o
diff --git a/drivers/power/bq2589x_chg_mmi/Kconfig b/drivers/power/bq2589x_chg_mmi/Kconfig
new file mode 100644
index 000000000000..96f25de40d2c
--- /dev/null
+++ b/drivers/power/bq2589x_chg_mmi/Kconfig
@@ -0,0 +1,31 @@
+#
+# BQ2589X Charger Driver
+#
+
+config BQ2589X_CHARGER
+    tristate "TI BQ2589x Charger Driver"
+    depends on MMI_DISCRETE_CHARGER_CLASS
+    default n
+    help
+        Say Y here if you have a bq2589x charging device in
+        your system.
+
+        To compile this driver as a module, choose M here: the
+        module will be called bq2589x_charger.
+
+config BQ2589X_MMI_QC3P_TURBO_CHARGER
+    bool "Enable BQ2589X Moto QC3P Turbo Charger"
+    depends on BQ2589X_CHARGER
+    default n
+    help
+        Say Y here if you have a BQ2589X Moto QC3P turbo charger compatible device connected.
+
+        If unsure, say N.
+
+config MMI_EXT_CHG_LED
+    bool "Enable MMI external charger LED"
+    default n
+    help
+        Say Y here if you have a MMI external charger LED.
+
+        If unsure, say N.
diff --git a/drivers/power/bq2597x_mmi/Kbuild b/drivers/power/bq2597x_mmi/Kbuild
index 8fbfa78a8359..26c3d62ca5ef 100644
--- a/drivers/power/bq2597x_mmi/Kbuild
+++ b/drivers/power/bq2597x_mmi/Kbuild
@@ -1,9 +1,4 @@
-# add -Wall to try to catch everything we can.
-EXTRA_CFLAGS += -Wall
-EXTRA_CFLAGS += -I$(TOP)/motorola/kernel/modules/include \
-		-I$(TOP)/motorola/kernel/modules/drivers/power/bq2597x_mmi
-
-obj-m += bq2597x_mmi.o
+obj-$(CONFIG_BQ2597X_MMI) += bq2597x_mmi.o
 
 ifneq ($(filter m y,$(CONFIG_BQ2597X_DUAL)),)
 	bq2597x_mmi-objs += bq25980_charger.o
diff --git a/drivers/power/bq2597x_mmi/Kconfig b/drivers/power/bq2597x_mmi/Kconfig
new file mode 100644
index 000000000000..8053a7ab9eb4
--- /dev/null
+++ b/drivers/power/bq2597x_mmi/Kconfig
@@ -0,0 +1,16 @@
+config BQ2597X_MMI
+	tristate "BQ2597X MMI"
+	default n
+	help
+	  Say Y here if you have a bq2597x compatible
+	  device connected.
+
+	  This driver can also be built as a module.  If so, the module
+	  will be called bq2597x_mmi.
+
+config BQ2597X_DUAL
+	bool "BQ2597X DUAL"
+	default n
+	help
+	  Say Y here if you have a bq2597x DUAL compatible
+	  device connected.
diff --git a/drivers/power/bq2597x_mmi_iio/Kbuild b/drivers/power/bq2597x_mmi_iio/Kbuild
index 4a5406b5538b..eaab1e1ea98c 100644
--- a/drivers/power/bq2597x_mmi_iio/Kbuild
+++ b/drivers/power/bq2597x_mmi_iio/Kbuild
@@ -1,7 +1 @@
-# add -Wall to try to catch everything we can.
-EXTRA_CFLAGS += -Wall
-EXTRA_CFLAGS += -I$(ANDROID_BUILD_TOP)/motorola/kernel/modules/include \
-		-I$(ANDROID_BUILD_TOP)/motorola/kernel/modules/drivers/power/bq2597x_mmi_iio
-
-obj-m += bq2597x_mmi_iio.o
-
+obj-$(CONFIG_BQ2597X_MMI_IIO) += bq2597x_mmi_iio.o
diff --git a/drivers/power/bq2597x_mmi_iio/Kconfig b/drivers/power/bq2597x_mmi_iio/Kconfig
new file mode 100644
index 000000000000..f27a0b63702d
--- /dev/null
+++ b/drivers/power/bq2597x_mmi_iio/Kconfig
@@ -0,0 +1,9 @@
+config BQ2597X_MMI_IIO
+	tristate "BQ2597X MMI IIO"
+	default n
+	help
+	  Say Y here if you have a bq2597x iio compatible
+	  device connected.
+
+	  This driver can also be built as a module.  If so, the module
+	  will be called bq2597x_mmi_iio.
diff --git a/drivers/power/cw2217b_fg_mmi/Kbuild b/drivers/power/cw2217b_fg_mmi/Kbuild
index a84af03167d3..b1dc7180229f 100644
--- a/drivers/power/cw2217b_fg_mmi/Kbuild
+++ b/drivers/power/cw2217b_fg_mmi/Kbuild
@@ -1,11 +1,6 @@
-# add -Wall to try to catch everything we can.
-EXTRA_CFLAGS += -Wall
-EXTRA_CFLAGS += -I$(ANDROID_BUILD_TOP)/motorola/kernel/modules/include
-EXTRA_CFLAGS += -I$(ANDROID_BUILD_TOP)/motorola/kernel/drivers/power/cw2217b_fg_mmi
-
 ifneq ($(filter m y,$(CONFIG_DYNAMIC_UPDATE_UI_FULL)),)
-	EXTRA_CFLAGS += -DDYNAMIC_UPDATE_UI_FULL
+	ccflags-y += -DDYNAMIC_UPDATE_UI_FULL
 endif
 
 cw2217b_fg_mmi-objs += cw2217B_fuel_gauge_base.o
-obj-m += cw2217b_fg_mmi.o
+obj-$(CONFIG_CW2217B_FG_MMI) += cw2217b_fg_mmi.o
diff --git a/drivers/power/cw2217b_fg_mmi/Kconfig b/drivers/power/cw2217b_fg_mmi/Kconfig
new file mode 100644
index 000000000000..0a512f41349b
--- /dev/null
+++ b/drivers/power/cw2217b_fg_mmi/Kconfig
@@ -0,0 +1,20 @@
+config CW2217B_FG_MMI
+    tristate "CW2217 FGADC Device Driver V1.2"
+    default n
+    help
+        Say Y here if you have CW2217 FGADC Device support
+        in your system.
+
+        If unsure, say N.
+
+        To compile this driver as a module, choose M here: the
+        module will be called cw2217b_fg_mmi.
+
+config DYNAMIC_UPDATE_UI_FULL
+    bool "Save the value of CW2217 ui_full to the 0xA2 register"
+    depends on CW2217B_FG_MMI
+    default n
+    help
+        Say Y here to save the value of ui_full to the 0xA2 register.
+
+        If unsure, say N.
diff --git a/drivers/power/mmi-smbcharger-iio/Kconfig b/drivers/power/mmi-smbcharger-iio/Kconfig
new file mode 100644
index 000000000000..60056fcf9782
--- /dev/null
+++ b/drivers/power/mmi-smbcharger-iio/Kconfig
@@ -0,0 +1,9 @@
+config MMI_SMBCHARGER_IIO
+        tristate "MMI SMBCHARGER IIO"
+        default n
+        help
+         Say Y here if you have a mmi smbcharger iio compatible
+         device connected.
+
+         This driver can also be built as a module.  If so, the module
+         will be called mmi-smbcharger-iio.
diff --git a/drivers/power/mmi_charger/Kbuild b/drivers/power/mmi_charger/Kbuild
index c1ab14765ca1..c372e5471463 100755
--- a/drivers/power/mmi_charger/Kbuild
+++ b/drivers/power/mmi_charger/Kbuild
@@ -1 +1 @@
-obj-$(CONFIG_USE_MMI_CHARGER) += mmi_charger.o
+obj-$(CONFIG_MMI_CHARGER) += mmi_charger.o
diff --git a/drivers/power/mmi_charger/Kconfig b/drivers/power/mmi_charger/Kconfig
index ad8495df665b..63aa01067e6f 100644
--- a/drivers/power/mmi_charger/Kconfig
+++ b/drivers/power/mmi_charger/Kconfig
@@ -1,9 +1,21 @@
-config USE_MMI_CHARGER
-        tristate "MMI CHARGER"
-        default n
-        help
-         Say Y here if you have a mmi charger compatible
-         device connected.
-
-         This driver can also be built as a module.  If so, the module
-         will be called mmi_charger.
+config MMI_CHARGER
+    tristate "MMI Charger Driver"
+    depends on MMI_INFO
+    default n
+    help
+        Say Y here if you have mmi charger support
+        in your system.
+
+        If unsure, say N.
+
+        To compile this driver as a module, choose M here: the
+        module will be called mmi_charger.
+
+config MMI_RECHARGER_HAWAO_MODE
+    bool "Enable MMI Recharger Hawao Mode"
+    depends on MMI_CHARGER
+    default n
+    help
+        Say Y here if you have a mmi recharger hawao mode compatible device connected.
+
+        If unsure, say N.
diff --git a/drivers/power/mmi_discrete_charger/Kbuild b/drivers/power/mmi_discrete_charger/Kbuild
index 7c149225341f..03d0b8d9cc7d 100644
--- a/drivers/power/mmi_discrete_charger/Kbuild
+++ b/drivers/power/mmi_discrete_charger/Kbuild
@@ -1,16 +1,10 @@
-# add -Wall to try to catch everything we can.
-EXTRA_CFLAGS += -Wall
-EXTRA_CFLAGS += -I$(ANDROID_BUILD_TOP)/motorola/kernel/modules/include
-EXTRA_CFLAGS +=	-I$(ANDROID_BUILD_TOP)/motorola/kernel/modules/drivers/power/mmi_charger
-EXTRA_CFLAGS +=	-I$(ANDROID_BUILD_TOP)/motorola/kernel/modules/drivers/power/mmi_discrete_charger
+ccflags-y += -I$(srctree)/drivers/power/mmi_charger
+ccflags-y += -I$(srctree)/drivers/power/mmi_discrete_charger
 
-obj-m += mmi_discrete_charger.o
+obj-$(CONFIG_MMI_DISCRETE_CHARGER) += mmi_discrete_charger.o
 mmi_discrete_charger-objs += mmi_discrete_charger_core.o
 mmi_discrete_charger-objs += mmi_discrete_charger_iio.o
 mmi_discrete_charger-objs += mmi_discrete_voter.o
 mmi_discrete_charger-objs += mmi_discrete_factory_tcmd.o
 
-KBUILD_EXTRA_SYMBOLS += $(CURDIR)/$(KBUILD_EXTMOD)/../../usb/typec/adapter_class/$(GKI_OBJ_MODULE_DIR)/Module.symvers
-KBUILD_EXTRA_SYMBOLS += $(CURDIR)/$(KBUILD_EXTMOD)/../mmi_charger/$(GKI_OBJ_MODULE_DIR)/Module.symvers
-
-obj-m += mmi_discrete_charger_class.o
+obj-$(CONFIG_MMI_DISCRETE_CHARGER_CLASS) += mmi_discrete_charger_class.o
diff --git a/drivers/power/mmi_discrete_charger/Kconfig b/drivers/power/mmi_discrete_charger/Kconfig
new file mode 100644
index 000000000000..7127dfae83fe
--- /dev/null
+++ b/drivers/power/mmi_discrete_charger/Kconfig
@@ -0,0 +1,24 @@
+config MMI_DISCRETE_CHARGER
+    tristate "MMI Discrete Charger Driver"
+    depends on TYPEC_ADAPTER_CLASS && MMI_CHARGER
+    default n
+    help
+        Say Y here if you have mmi discrete charger support
+        in your system.
+
+        If unsure, say N.
+
+        To compile this driver as a module, choose M here: the
+        module will be called mmi_discrete_charger.
+
+config MMI_DISCRETE_CHARGER_CLASS
+    tristate "MMI Discrete Charger Class Device"
+    default n
+    help
+        Say Y here if you have mmi discrete charger class support
+        in your system.
+
+        If unsure, say N.
+
+        To compile this driver as a module, choose M here: the
+        module will be called mmi_discrete_charger_class.
diff --git a/drivers/power/mmi_discrete_turbo_charger/Kbuild b/drivers/power/mmi_discrete_turbo_charger/Kbuild
index b6d3765de57c..415f13d845d3 100644
--- a/drivers/power/mmi_discrete_turbo_charger/Kbuild
+++ b/drivers/power/mmi_discrete_turbo_charger/Kbuild
@@ -1,13 +1,4 @@
-# add -Wall to try to catch everything we can.
-EXTRA_CFLAGS += -Wall
-EXTRA_CFLAGS += -I$(ANDROID_BUILD_TOP)/motorola/kernel/modules/include \
-		-I$(ANDROID_BUILD_TOP)/motorola/kernel/modules/drivers/power/mmi_discrete_turbo_charger
-
-ifneq ($(filter m y,$(CONFIG_MMI_QC3P_WT6670_DETECTED)),)
-        EXTRA_CFLAGS += -DCONFIG_MMI_QC3P_WT6670_DETECTED
-endif
-
-obj-m += mmi_discrete_turbo_charger.o
+obj-$(CONFIG_MMI_DISCRETE_TURBO_CHARGER) += mmi_discrete_turbo_charger.o
 mmi_discrete_turbo_charger-objs += mmi_charger_class.o
 mmi_discrete_turbo_charger-objs += mmi_charger_core.o
 mmi_discrete_turbo_charger-objs += mmi_charger_pump_policy.o
@@ -15,8 +6,3 @@ mmi_discrete_turbo_charger-objs += mmi_cp_charger.o
 mmi_discrete_turbo_charger-objs += qpnp_pmic_charger.o
 mmi_discrete_turbo_charger-objs += mmi_qc3p.o
 mmi_discrete_turbo_charger-objs += mmi_qc3p_cp_policy.o
-
-#ifeq ($(CONFIG_MMI_PL_CP_POLICY), y)
-#endif
-
-KBUILD_EXTRA_SYMBOLS += $(CURDIR)/$(KBUILD_EXTMOD)/../../usb/typec/adapter_class/$(GKI_OBJ_MODULE_DIR)/Module.symvers
diff --git a/drivers/power/mmi_discrete_turbo_charger/Kconfig b/drivers/power/mmi_discrete_turbo_charger/Kconfig
new file mode 100644
index 000000000000..f3cb071c55e5
--- /dev/null
+++ b/drivers/power/mmi_discrete_turbo_charger/Kconfig
@@ -0,0 +1,12 @@
+config MMI_DISCRETE_TURBO_CHARGER
+    tristate "MMI Discrete Turbo Charger Driver"
+    depends on TYPEC_ADAPTER_CLASS && MMI_CHARGER
+    default n
+    help
+        Say Y here if you have mmi discrete turbo charger support
+        in your system.
+
+        If unsure, say N.
+
+        To compile this driver as a module, choose M here: the
+        module will be called mmi_discrete_turbo_charger.
diff --git a/drivers/power/mmi_parallel_charger_iio/Kconfig b/drivers/power/mmi_parallel_charger_iio/Kconfig
new file mode 100644
index 000000000000..15dc8d0eb981
--- /dev/null
+++ b/drivers/power/mmi_parallel_charger_iio/Kconfig
@@ -0,0 +1,9 @@
+config MMI_PARALLEL_CHARGER_IIO
+        tristate "MMI Parallel Charger IIO"
+        default n
+        help
+         Say Y here if you have a mmi parallel charger iio compatible
+         device connected.
+
+         This driver can also be built as a module.  If so, the module
+         will be called mmi_parallel_charger_iio.
diff --git a/drivers/power/mmi_sc8549/Kconfig b/drivers/power/mmi_sc8549/Kconfig
new file mode 100644
index 000000000000..8933e161024f
--- /dev/null
+++ b/drivers/power/mmi_sc8549/Kconfig
@@ -0,0 +1,9 @@
+config MMI_SC8549
+        tristate "MMI SC8549 PD"
+        default n
+        help
+         Say Y here if you have a mmi sc8549 compatible
+         device connected.
+
+         This driver can also be built as a module.  If so, the module
+         will be called mmi_sc8549.
diff --git a/drivers/power/qpnp_adaptive_charge/Kbuild b/drivers/power/qpnp_adaptive_charge/Kbuild
index 41d52ec8b3d5..ed4c6c7a9d5a 100755
--- a/drivers/power/qpnp_adaptive_charge/Kbuild
+++ b/drivers/power/qpnp_adaptive_charge/Kbuild
@@ -1,6 +1,6 @@
-ifneq ($(filter m y,$(CONFIG_USE_MMI_CHARGER)),)
-	EXTRA_CFLAGS += -DUSE_MMI_CHARGER
-	EXTRA_CFLAGS += -I$(srctree)/$(src)/../mmi_charger
+ifneq ($(filter m y,$(CONFIG_MMI_CHARGER)),)
+	ccflags-y += -DUSE_MMI_CHARGER
+	ccflags-y += -I$(srctree)/$(src)/../mmi_charger
 endif
 
 ifneq ($(filter m y, $(ADAPTIVE_TOLERANCE_OPTIMIZATION)),)
diff --git a/drivers/power/rt9426a_fg_mmi/Kconfig b/drivers/power/rt9426a_fg_mmi/Kconfig
new file mode 100644
index 000000000000..64809be4bde9
--- /dev/null
+++ b/drivers/power/rt9426a_fg_mmi/Kconfig
@@ -0,0 +1,9 @@
+config RT9426A_FG
+        tristate "RT9426A Fuel Guage driver"
+        default n
+        help
+         Say Y here if you have a RT9426A compatible
+         device connected.
+
+         This driver can also be built as a module.  If so, the module
+         will be called rt9426a_fg_mmi.
diff --git a/drivers/power/sd77426_fg_mmi/Kconfig b/drivers/power/sd77426_fg_mmi/Kconfig
new file mode 100644
index 000000000000..dbaf0ad0d216
--- /dev/null
+++ b/drivers/power/sd77426_fg_mmi/Kconfig
@@ -0,0 +1,9 @@
+config SD77426_FG
+        tristate "SD77426 Fuel Guage driver"
+        default n
+        help
+         Say Y here if you have a SD77426 compatible
+         device connected.
+
+         This driver can also be built as a module.  If so, the module
+         will be called sd77426_fg_mmi.
diff --git a/drivers/power/sgm4154x_chg_mmi/Kbuild b/drivers/power/sgm4154x_chg_mmi/Kbuild
index b9fd084d799a..e1f8c8b580f0 100644
--- a/drivers/power/sgm4154x_chg_mmi/Kbuild
+++ b/drivers/power/sgm4154x_chg_mmi/Kbuild
@@ -1,27 +1,5 @@
-# add -Wall to try to catch everything we can.
-EXTRA_CFLAGS += -Wall
-EXTRA_CFLAGS += -I$(ANDROID_BUILD_TOP)/motorola/kernel/modules/include \
-		-I$(ANDROID_BUILD_TOP)/motorola/kernel/modules/drivers/power/sgm4154x_chg_mmi
-
 ifneq ($(filter m y,$(CONFIG_MMI_QC3P_TURBO_CHARGER)),)
-        EXTRA_CFLAGS += -DCONFIG_MMI_QC3P_TURBO_CHARGER
-endif
-
-ifneq ($(filter m y,$(CONFIG_MMI_QC3P_WT6670_DETECTED)),)
-        EXTRA_CFLAGS += -DCONFIG_MMI_QC3P_WT6670_DETECTED
-endif
-
-ifneq ($(filter m y,$(CONFIG_MMI_SGM41513_CHARGER)),)
-        EXTRA_CFLAGS += -DCONFIG_MMI_SGM41513_CHARGER
-endif
-
-ifneq ($(filter m y,$(CONFIG_MMI_QC3P_Z350_DETECTED)),)
-        EXTRA_CFLAGS += -DCONFIG_MMI_QC3P_Z350_DETECTED
-endif
-
-ifneq ($(filter m y,$(CONFIG_MMI_EXT_CHG_LED)),)
-        EXTRA_CFLAGS += -DCONFIG_MMI_EXT_CHG_LED
+    ccflags-y += -DCONFIG_MMI_QC3P_TURBO_CHARGER
 endif
 
-obj-m += sgm4154x_charger.o
-KBUILD_EXTRA_SYMBOLS += $(CURDIR)/$(KBUILD_EXTMOD)/../mmi_discrete_charger/$(GKI_OBJ_MODULE_DIR)/Module.symvers
+obj-$(CONFIG_SGM4154X_CHARGER) += sgm4154x_charger.o
diff --git a/drivers/power/sgm4154x_chg_mmi/Kconfig b/drivers/power/sgm4154x_chg_mmi/Kconfig
new file mode 100644
index 000000000000..c34882898d96
--- /dev/null
+++ b/drivers/power/sgm4154x_chg_mmi/Kconfig
@@ -0,0 +1,23 @@
+#
+# SGM4154X Charger driver
+#
+
+config SGM4154X_CHARGER
+    tristate "sgm4154x charger driver"
+    depends on MMI_DISCRETE_CHARGER_CLASS
+    default n
+    help
+        Say Y here if you have a sgm415x charging device in
+        your system.
+
+        To compile this driver as a module, choose M here: the
+        module will be called sgm415x_charger.
+
+config MMI_SGM41513_CHARGER
+    bool "Enable SGM4154X Moto SGM41513 Charger"
+    depends on SGM4154X_CHARGER
+    default n
+    help
+        Say Y here if you have a SGM4154X Moto SGM41513 Charger compatible device connected.
+
+        If unsure, say N.
diff --git a/drivers/power/sm5602_fg_mmi/Kconfig b/drivers/power/sm5602_fg_mmi/Kconfig
new file mode 100644
index 000000000000..56a33f5f73b8
--- /dev/null
+++ b/drivers/power/sm5602_fg_mmi/Kconfig
@@ -0,0 +1,3 @@
+config MMI_SM5602_FG
+        tristate "sm5602_fg_mmi"
+        default n
diff --git a/drivers/regulator/Kconfig b/drivers/regulator/Kconfig
index df0499425657..fc509b133964 100644
--- a/drivers/regulator/Kconfig
+++ b/drivers/regulator/Kconfig
@@ -1232,5 +1232,9 @@ config VIRTIO_REGULATOR
 
 source "drivers/regulator/wl2864c/Kconfig"
 
+source "drivers/regulator/wl2866d/Kconfig"
+
+source "drivers/regulator/wl2868c/Kconfig"
+
 endif
 
diff --git a/drivers/regulator/Makefile b/drivers/regulator/Makefile
index 1d4048de63fd..6f26ec7de99f 100644
--- a/drivers/regulator/Makefile
+++ b/drivers/regulator/Makefile
@@ -152,5 +152,7 @@ obj-$(CONFIG_REGULATOR_STUB) += stub-regulator.o
 obj-$(CONFIG_VIRTIO_REGULATOR) += virtio_regulator.o
 
 obj-$(CONFIG_REGULATOR_WL2864C) += wl2864c/
+obj-$(CONFIG_REGULATOR_WL2866D) += wl2866d/
+obj-$(CONFIG_REGULATOR_WL2868C) += wl2868c/
 
 ccflags-$(CONFIG_REGULATOR_DEBUG) += -DDEBUG
diff --git a/drivers/regulator/wl2866d/Kconfig b/drivers/regulator/wl2866d/Kconfig
new file mode 100644
index 000000000000..3a713b20a0b4
--- /dev/null
+++ b/drivers/regulator/wl2866d/Kconfig
@@ -0,0 +1,9 @@
+config REGULATOR_WL2866D
+        tristate "WL2866D REGULATOR"
+        default n
+        help
+         Say Y here if you have a WL2866D compatible
+         device connected.
+
+         This driver can also be built as a module.  If so, the module
+         will be called wl2866d-regulator.
diff --git a/drivers/regulator/wl2868c/Kconfig b/drivers/regulator/wl2868c/Kconfig
new file mode 100644
index 000000000000..7e6c4f2049f0
--- /dev/null
+++ b/drivers/regulator/wl2868c/Kconfig
@@ -0,0 +1,9 @@
+config REGULATOR_WL2868C
+        tristate "WL2868C REGULATOR"
+        default n
+        help
+         Say Y here if you have a WL2868C compatible
+         device connected.
+
+         This driver can also be built as a module.  If so, the module
+         will be called wl2868c-regulator.
diff --git a/drivers/usb/typec/Kconfig b/drivers/usb/typec/Kconfig
index 895e2418de53..07f924033362 100644
--- a/drivers/usb/typec/Kconfig
+++ b/drivers/usb/typec/Kconfig
@@ -50,6 +50,8 @@ source "drivers/usb/typec/tcpm/Kconfig"
 
 source "drivers/usb/typec/ucsi/Kconfig"
 
+source "drivers/usb/typec/mmi_tcpc/Kconfig"
+
 config TYPEC_TPS6598X
 	tristate "TI TPS6598x USB Power Delivery controller driver"
 	depends on I2C
@@ -65,4 +67,6 @@ source "drivers/usb/typec/mux/Kconfig"
 
 source "drivers/usb/typec/altmodes/Kconfig"
 
+source "drivers/usb/typec/adapter_class/Kconfig"
+
 endif # TYPEC
diff --git a/drivers/usb/typec/Makefile b/drivers/usb/typec/Makefile
index 6696b7263d61..6a5b945408d7 100644
--- a/drivers/usb/typec/Makefile
+++ b/drivers/usb/typec/Makefile
@@ -6,3 +6,5 @@ obj-$(CONFIG_TYPEC_TCPM)	+= tcpm/
 obj-$(CONFIG_TYPEC_UCSI)	+= ucsi/
 obj-$(CONFIG_TYPEC_TPS6598X)	+= tps6598x.o
 obj-$(CONFIG_TYPEC)		+= mux/
+obj-$(CONFIG_TCPC_CLASS)	+= mmi_tcpc/
+obj-$(CONFIG_TYPEC_ADAPTER_CLASS) += adapter_class/
diff --git a/drivers/usb/typec/adapter_class/Kbuild b/drivers/usb/typec/adapter_class/Kbuild
index f003c3c245e9..9e88ccb1f252 100755
--- a/drivers/usb/typec/adapter_class/Kbuild
+++ b/drivers/usb/typec/adapter_class/Kbuild
@@ -1,5 +1 @@
-# add -Wall to try to catch everything we can.
-EXTRA_CFLAGS += -Wall
-EXTRA_CFLAGS += -I$(ANDROID_BUILD_TOP)/motorola/kernel/modules/include
-
-obj-m += adapter_class.o
+obj-$(CONFIG_TYPEC_ADAPTER_CLASS) += adapter_class.o
diff --git a/drivers/usb/typec/adapter_class/Kconfig b/drivers/usb/typec/adapter_class/Kconfig
new file mode 100644
index 000000000000..e61abcc415e3
--- /dev/null
+++ b/drivers/usb/typec/adapter_class/Kconfig
@@ -0,0 +1,11 @@
+config TYPEC_ADAPTER_CLASS
+    tristate "Adapter Class Device Driver"
+    default n
+    help
+        Say Y here if you have Adapter Class Device support
+        in your system.
+
+        If unsure, say N.
+
+        To compile this driver as a module, choose M here: the
+        module will be called adapter_class.
diff --git a/drivers/usb/typec/mmi_tcpc/Kbuild b/drivers/usb/typec/mmi_tcpc/Kbuild
index 72f7413ba5bc..a9165010cf72 100644
--- a/drivers/usb/typec/mmi_tcpc/Kbuild
+++ b/drivers/usb/typec/mmi_tcpc/Kbuild
@@ -1,18 +1,13 @@
-# add -Wall to try to catch everything we can.
-EXTRA_CFLAGS += -Wall
-EXTRA_CFLAGS += -I$(ANDROID_BUILD_TOP)/motorola/kernel/modules/include \
-		-I$(ANDROID_BUILD_TOP)/motorola/kernel/modules/drivers/usb/typec/mmi_tcpc
-
 subdir-ccflags-y += -Wall -Werror -DCONFIG_RT_REGMAP
 
-EXTRA_CFLAGS += -DCONFIG_PD_DBG_INFO
+ccflags-y += -DCONFIG_PD_DBG_INFO
 
 ifneq ($(filter m y,$(CONFIG_TCPC_CLASS)),)
-	EXTRA_CFLAGS += -DCONFIG_TCPC_CLASS
+	ccflags-y += -DCONFIG_TCPC_CLASS
 endif
 
 ifneq ($(filter m y,$(CONFIG_USB_POWER_DELIVERY)),)
-	EXTRA_CFLAGS += -DCONFIG_USB_POWER_DELIVERY
+	ccflags-y += -DCONFIG_USB_POWER_DELIVERY
 endif
 
 ifneq ($(filter m y,$(CONFIG_TCPC_RT1711H)),)
@@ -28,7 +23,7 @@ ifneq ($(filter m y,$(CONFIG_TCPC_SGM7220)),)
 endif
 
 ifneq ($(CONFIG_TCPC_MAX_POLLING_COUNT),)
-        EXTRA_CFLAGS += -DCONFIG_TCPC_MAX_POLLING_COUNT=$(CONFIG_TCPC_MAX_POLLING_COUNT)
+        ccflags-y += -DCONFIG_TCPC_MAX_POLLING_COUNT=$(CONFIG_TCPC_MAX_POLLING_COUNT)
 endif
 
 tcpc_class-objs	+= tcpci_core.o tcpci_typec.o tcpci_timer.o \
@@ -48,8 +43,5 @@ tcpc_class-objs	+= tcpci_event.o \
 		pd_dpm_mmi_adapter.o
 endif
 
-obj-m += tcpc_class.o
-obj-m += rt_pd_manager.o
-
-KBUILD_EXTRA_SYMBOLS += $(CURDIR)/$(KBUILD_EXTMOD)/../adapter_class/$(GKI_OBJ_MODULE_DIR)/Module.symvers
-KBUILD_EXTRA_SYMBOLS += $(CURDIR)/$(KBUILD_EXTMOD)/../../../power/mmi_discrete_charger/$(GKI_OBJ_MODULE_DIR)/Module.symvers
+obj-$(CONFIG_TCPC_CLASS) += tcpc_class.o
+obj-$(CONFIG_RT_PD_MANAGER) += rt_pd_manager.o
diff --git a/drivers/usb/typec/mmi_tcpc/Kconfig b/drivers/usb/typec/mmi_tcpc/Kconfig
new file mode 100644
index 000000000000..e6f01faf331f
--- /dev/null
+++ b/drivers/usb/typec/mmi_tcpc/Kconfig
@@ -0,0 +1,50 @@
+config TCPC_RT1711H
+        tristate "TCPC RT1711H driver"
+        default n
+        depends on TCPC_CLASS
+        help
+         Say Y here if you have a rt1711h compatible
+         device connected.
+
+         This driver can also be built as a module.  If so, the module
+         will be called tcpc_rt1711h.
+
+config TCPC_SGM7220
+        tristate "TCPC SGM7220 driver"
+        default n
+        depends on TCPC_CLASS
+        help
+         Say Y here if you have a sgm7220 compatible
+         device connected.
+
+         This driver can also be built as a module.  If so, the module
+         will be called tcpc_sgm7220.
+
+config TCPC_CLASS
+        tristate "TCPC Class driver"
+        default n
+        help
+         Say Y here if you have a tcpc compatible
+         device connected.
+
+         This driver can also be built as a module.  If so, the module
+         will be called tcpc_class.
+
+config RT_PD_MANAGER
+        tristate "RT PD Manager driver"
+        default n
+        help
+         Say Y here if you have a RT PD compatible
+         device connected.
+
+         This driver can also be built as a module.  If so, the module
+         will be called rt_pd_manager.
+
+config USB_POWER_DELIVERY
+        bool "Enable USB Power Delivery"
+        depends on TCPC_CLASS && TYPEC_ADAPTER_CLASS
+        default n
+        help
+         Say Y here if you have a usb power delivery compatible device connected.
+
+         If unsure, say N.
-- 
2.39.5

